mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -511,30 +511,44 @@ where
     T: Read + Seek,
 {
     // We don't need any explicit returns elsewhere as EOF should be an ERR.
+
+    // There shouldn't be any issues from cast as block size is less than u64 max and usize max always.
     #[allow(clippy::cast_possible_truncation)]
-    let all_zeroes: Vec<u8> = vec![0; *SERIALIZED_BLOCK_SIZE as usize];
-    loop {
-        #[allow(clippy::cast_possible_truncation)]
-        let mut buf = vec![0; *SERIALIZED_BLOCK_SIZE as usize];
-        readable.read_exact(&mut buf)?;
+    let serialized_block_size = *SERIALIZED_BLOCK_SIZE as usize;
+    let buffer_size = (serialized_block_size * 2) as usize;
 
-        // If all zeroes we can skip faster.
+    // Read two blocks worth of data and scan to see if there is a block.
+    // If not, then read another block worth of data and continue...
+    let mut buf = vec![0; buffer_size];
+    readable.read_exact(&mut buf)?;
+
+    let all_zeroes = vec![0; buffer_size];
+    loop {
         if buf == all_zeroes {
-            // This is okay since the block size is too small to wrap.
-            #[allow(clippy::cast_possible_wrap)]
-            readable.seek(SeekFrom::Current(*SERIALIZED_BLOCK_SIZE as i64))?;
+            readable.read_exact(&mut buf)?;
             continue;
         }
 
-        if let Ok(block) = bincode::deserialize::<BlockHeaderWithCrc>(&buf) {
-            let BlockVersion::Version1(inner) = block.inner();
-            // We maybe found a block.
-            if inner.hint() == BLOCK_HINT {
-                return Ok(block);
+        // Try to find block in the buffer we have, if cannot find it by halfway point,
+        // we will need to fetch more data.
+        for offset in 0..(serialized_block_size) {
+            let start = offset;
+            let end = start + serialized_block_size;
+            if let Ok(block) = bincode::deserialize::<BlockHeaderWithCrc>(&buf[start..end]) {
+                let BlockVersion::Version1(inner) = block.inner();
+                // We maybe found a block.
+                if inner.hint() == BLOCK_HINT {
+                    return Ok(block);
+                }
             }
         }
 
-        readable.seek(SeekFrom::Current(1))?;
+        // Remove bytes we know don't have a block in them.
+        buf.drain(..serialized_block_size);
+        // extend buffer to fit another potential blocks worth of data
+        buf.resize(buffer_size, 0);
+        // Read into second half of the buffer
+        readable.read_exact(&mut buf[serialized_block_size..])?;
     }
 }
```

Comment:
Makes sense :+1:. If this can never find a block then `read_exact()` will error eventually right?

Reply:
Yep, that is expected. In the `find_pointers_and_order_post_crash` fn, there is a check for any Err. If there is one, we return a default RingBuffer.